graph_dfs_debug/graph.py

```diff
@@ -19,36 +19,40 @@ def add_vertex(self, vertex, edges=()):
         self.vertices[vertex] = set(edges)
 
     def add_edge(self, start, end, bidirectional=True):
-        self.vertices[start].add(start)
+        self.vertices[start].add(end)
         if bidirectional:
-            self.vertices[end].add(end)
+            self.vertices[end].add(start)
 
     def dfs(self, start, target=None):
-        x = []
-        x.append(start)
-        y = set(x)
-
-        while x:
-            z = x.pop()
-            if x == target:
-                break
-            x.extend(self.vertices[z])
-
-        return x
-
-    def graph_rec(self, start, target=None):
-        x = set()
-        x.append(start)
-        for v in self.vertices[start]:
-            graph_rec(v)
-        return x
+        stack = []
+        stack.append(start)
+        explored = set()
+
+        while stack:
+            vertex = stack.pop()
+            if vertex not in explored:
+                explored.add(vertex)
+                if vertex == target:
+                    break
+                stack.extend(self.vertices[vertex])
+        return explored
+
```

Comment:
Might want to add an explored.update here with the current vertex so that you update that set